README.md
# fnes


ChangeLog.md
# Changelog for fnes

## Unreleased changes


test/CPU/ADCSpec.hs
module CPU.ADCSpec (spec) where

import Test.Hspec
import CPU6502
import Memory

spec :: Spec
spec = do
    describe "IMM" $ do
        let cpu = iterateProcessor $ newCPU 
                $ (0x8000 <-> [0x69, 0x42]) -- ADC #$42
        it "check PC register" $ do (getPCRegister cpu) `shouldBe` (0x8000 + 2)
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x42
        it "check ticks" $ do (tick cpu) `shouldBe` 2
    describe "ZP" $ do
        let cpu = iterateProcessor $ newCPU 
                $ (0x0000 <-> [0x42])
                . (0x8000 <-> [0x65, 0x00]) -- ADC *00
        it "check PC register" $ do (getPCRegister cpu) `shouldBe` (0x8000 + 2)
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x42
        it "check ticks" $ do (tick cpu) `shouldBe` 3
    describe "ZPX" $ do
        let cpu = iterateProcessor 
                $ setXRegister 1 
                $ newCPU 
                $ (0x0000 <-> [0x00, 0x42])
                . (0x8000 <-> [0x75, 0x00]) -- ADC *00,X
        it "check PC register" $ do (getPCRegister cpu) `shouldBe` (0x8000 + 2)
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x42
        it "check ticks" $ do (tick cpu) `shouldBe` 4
    describe "ABS" $ do
        let cpu = iterateProcessor $ newCPU 
                $ (0x7900 <-> [0x42])
                . (0x8000 <-> [0x6D, 0x00, 0x79]) -- ADC 7900
        it "check PC register" $ do (getPCRegister cpu) `shouldBe` (0x8000 + 3)
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x42
        it "check ticks" $ do (tick cpu) `shouldBe` 4
    describe "ABSX" $ do
        let cpu = iterateProcessor $ setXRegister 1 $ newCPU 
                $ (0x7900 <-> [0x00, 0x42])
                . (0x8000 <-> [0x7D, 0x00, 0x79]) -- ADC 7900,X
        it "check PC register" $ do (getPCRegister cpu) `shouldBe` (0x8000 + 3)
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x42
        it "check ticks" $ do (tick cpu) `shouldBe` 4
    describe "ABSY" $ do
        let cpu = iterateProcessor 
                $ setYRegister 1 
                $ newCPU 
                $ (0x7900 <-> [0x00, 0x42])
                . (0x8000 <-> [0x79, 0x00, 0x79]) -- ADC 7900,Y
        it "check PC register" $ do (getPCRegister cpu) `shouldBe` (0x8000 + 3)
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x42
        it "check ticks" $ do (tick cpu) `shouldBe` 4
    describe "INDX" $ do
        let cpu = iterateProcessor 
                $ setXRegister 1 
                $ newCPU 
                $ (0x0000 <-> [0x00, 0x00, 0x01, 0x79])
                . (0x7900 <-> [0x00, 0x42])
                . (0x8000 <-> [0x61, 0x01]) -- ADC (1,X)
        it "check PC register" $ do (getPCRegister cpu) `shouldBe` (0x8000 + 2)
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x42
        it "check ticks" $ do (tick cpu) `shouldBe` 6 
    describe "INDY" $ do
        let cpu = iterateProcessor 
                $ setYRegister 1 
                $ newCPU 
                $ (0x0000 <-> [0x00, 0x00, 0x79])
                . (0x7900 <-> [0x00, 0x42])
                . (0x8000 <-> [0x71, 0x01]) -- ADC (1),Y
        it "check PC register" $ do (getPCRegister cpu) `shouldBe` (0x8000 + 2)
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x42
        it "check ticks" $ do (tick cpu) `shouldBe` 5 
    describe "0+0" $ do
        let cpu = iterateProcessor 
                $ setAccRegister 0x00
                $ setFlagC False 
                $ newCPU 
                $ (0x8000 <-> [0x69, 0x00])
        it "check flag C" $ do (getFlagC cpu) `shouldBe` False
        it "check flag Z" $ do (getFlagZ cpu) `shouldBe` True
        it "check flag N" $ do (getFlagN cpu) `shouldBe` False
        it "check flag V" $ do (getFlagV cpu) `shouldBe` False
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x00
    describe "0+0 with carry" $ do
        let cpu = iterateProcessor 
                $ setAccRegister 0x00
                $ setFlagC True 
                $ newCPU 
                $ (0x8000 <-> [0x69, 0x00])
        it "check flag C" $ do (getFlagC cpu) `shouldBe` False
        it "check flag Z" $ do (getFlagZ cpu) `shouldBe` False
        it "check flag N" $ do (getFlagN cpu) `shouldBe` False
        it "check flag V" $ do (getFlagV cpu) `shouldBe` False
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x01
    describe "2+2" $ do
        let cpu = iterateProcessor 
                $ setAccRegister 0x02
                $ setFlagC False 
                $ newCPU 
                $ (0x8000 <-> [0x69, 0x02])
        it "check flag C" $ do (getFlagC cpu) `shouldBe` False
        it "check flag Z" $ do (getFlagZ cpu) `shouldBe` False
        it "check flag N" $ do (getFlagN cpu) `shouldBe` False
        it "check flag V" $ do (getFlagV cpu) `shouldBe` False
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x04
    describe "2+2 with carry" $ do
        let cpu = iterateProcessor 
                $ setAccRegister 0x02
                $ setFlagC True 
                $ newCPU 
                $ (0x8000 <-> [0x69, 0x02])
        it "check flag C" $ do (getFlagC cpu) `shouldBe` False
        it "check flag Z" $ do (getFlagZ cpu) `shouldBe` False
        it "check flag N" $ do (getFlagN cpu) `shouldBe` False
        it "check flag V" $ do (getFlagV cpu) `shouldBe` False
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x05
    describe "0x01+0xFF" $ do
        let cpu = iterateProcessor 
                $ setAccRegister 0x01
                $ setFlagC False
                $ newCPU 
                $ (0x8000 <-> [0x69, 0xFF])
        it "check flag C" $ do (getFlagC cpu) `shouldBe` True
        it "check flag Z" $ do (getFlagZ cpu) `shouldBe` True
        it "check flag N" $ do (getFlagN cpu) `shouldBe` False
        it "check flag V" $ do (getFlagV cpu) `shouldBe` False
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x00
    describe "0x00+0xFF with carry" $ do
        let cpu = iterateProcessor 
                $ setAccRegister 0x00
                $ setFlagC True
                $ newCPU 
                $ (0x8000 <-> [0x69, 0xFF])
        it "check flag C" $ do (getFlagC cpu) `shouldBe` True
        it "check flag Z" $ do (getFlagZ cpu) `shouldBe` True
        it "check flag N" $ do (getFlagN cpu) `shouldBe` False
        it "check flag V" $ do (getFlagV cpu) `shouldBe` False
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x00
    describe "0x00+0x7F" $ do
        let cpu = iterateProcessor 
                $ setAccRegister 0x00
                $ setFlagC False
                $ newCPU 
                $ (0x8000 <-> [0x69, 0x7F])
        it "check flag C" $ do (getFlagC cpu) `shouldBe` False
        it "check flag Z" $ do (getFlagZ cpu) `shouldBe` False
        it "check flag N" $ do (getFlagN cpu) `shouldBe` False
        it "check flag V" $ do (getFlagV cpu) `shouldBe` False
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x7F
    describe "0x00+0x7F with carry" $ do
        let cpu = iterateProcessor 
                $ setAccRegister 0x00
                $ setFlagC True 
                $ newCPU 
                $ (0x8000 <-> [0x69, 0x7F])
        it "check flag C" $ do (getFlagC cpu) `shouldBe` False
        it "check flag Z" $ do (getFlagZ cpu) `shouldBe` False
        it "check flag N" $ do (getFlagN cpu) `shouldBe` True 
        it "check flag V" $ do (getFlagV cpu) `shouldBe` True 
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x80
    describe "0x01+0x7F" $ do
        let cpu = iterateProcessor 
                $ setAccRegister 0x01
                $ setFlagC False
                $ newCPU 
                $ (0x8000 <-> [0x69, 0x7F])
        it "check flag C" $ do (getFlagC cpu) `shouldBe` False
        it "check flag Z" $ do (getFlagZ cpu) `shouldBe` False
        it "check flag N" $ do (getFlagN cpu) `shouldBe` True 
        it "check flag V" $ do (getFlagV cpu) `shouldBe` True 
        it "check ACC register" $ do (getAccRegister cpu) `shouldBe` 0x80




test/CPUCommandsSpec.hs
module CPUCommandsSpec (spec) where

import Test.Hspec
import CPU6502
import Memory

spec :: Spec
spec = do
    describe "ADC" $ do
        it "check IMM" $ do
            let cpu = iterateProcessor $ newCPU 
                    $ (0x8000 <-> [0x69, 0x42]) -- ADC #$42
            (getPCRegister cpu) `shouldBe` (0x8000 + 2)
            (getAccRegister cpu) `shouldBe` 0x42
            (tick cpu) `shouldBe` 2
        it "check ZP" $ do
            let cpu = iterateProcessor $ newCPU 
                    $ (0x0000 <-> [0x42])
                    . (0x8000 <-> [0x65, 0x00]) -- ADC *00
            (getPCRegister cpu) `shouldBe` (0x8000 + 2)
            (getAccRegister cpu) `shouldBe` 0x42
            (tick cpu) `shouldBe` 3
        it "check ZPX" $ do
            let cpu = iterateProcessor 
                    $ setXRegister 1 
                    $ newCPU 
                    $ (0x0000 <-> [0x00, 0x42])
                    . (0x8000 <-> [0x75, 0x00]) -- ADC *00,X
            (getPCRegister cpu) `shouldBe` (0x8000 + 2)
            (getAccRegister cpu) `shouldBe` 0x42
            (tick cpu) `shouldBe` 4
        it "check ABS" $ do
            let cpu = iterateProcessor $ newCPU 
                    $ (0x7900 <-> [0x42])
                    . (0x8000 <-> [0x6D, 0x00, 0x79]) -- ADC 7900
            (getPCRegister cpu) `shouldBe` (0x8000 + 3)
            (getAccRegister cpu) `shouldBe` 0x42
            (tick cpu) `shouldBe` 4
        it "check ABSX" $ do
            let cpu = iterateProcessor $ setXRegister 1 $ newCPU 
                    $ (0x7900 <-> [0x00, 0x42])
                    . (0x8000 <-> [0x7D, 0x00, 0x79]) -- ADC 7900,X
            (getPCRegister cpu) `shouldBe` (0x8000 + 3)
            (getAccRegister cpu) `shouldBe` 0x42
            (tick cpu) `shouldBe` 4
        it "check ABSY" $ do
            let cpu = iterateProcessor 
                    $ setYRegister 1 
                    $ newCPU 
                    $ (0x7900 <-> [0x00, 0x42])
                    . (0x8000 <-> [0x79, 0x00, 0x79]) -- ADC 7900,Y
            (getPCRegister cpu) `shouldBe` (0x8000 + 3)
            (getAccRegister cpu) `shouldBe` 0x42
            (tick cpu) `shouldBe` 4
        it "check INDX" $ do
            let cpu = iterateProcessor 
                    $ setXRegister 1 
                    $ newCPU 
                    $ (0x0000 <-> [0x00, 0x00, 0x01, 0x79])
                    . (0x7900 <-> [0x00, 0x42])
                    . (0x8000 <-> [0x61, 0x01]) -- ADC (1,X)
            (getPCRegister cpu) `shouldBe` (0x8000 + 2)
            (getAccRegister cpu) `shouldBe` 0x42
            (tick cpu) `shouldBe` 6 
        it "check INDY" $ do
            let cpu = iterateProcessor 
                    $ setYRegister 1 
                    $ newCPU 
                    $ (0x0000 <-> [0x00, 0x00, 0x79])
                    . (0x7900 <-> [0x00, 0x42])
                    . (0x8000 <-> [0x71, 0x01]) -- ADC (1),Y
            (getPCRegister cpu) `shouldBe` (0x8000 + 2)
            (getAccRegister cpu) `shouldBe` 0x42
            (tick cpu) `shouldBe` 5 




src/Memory.hs
{-# LANGUAGE TypeSynonymInstances #-}
{-# LANGUAGE FlexibleInstances #-}

module Memory where

import System.Environment
import Data.Bits


class Memory a where
    read8 :: Int -> a -> Int
    read16 :: Int -> a -> (Int, Int)
    write8 :: Int -> Int -> a -> a
    write16 :: Int -> (Int, Int) -> a -> a

type ListMemory = [[Int]] 

type MemoryBuilder = ListMemory -> ListMemory 

(<->) :: Int -> [Int] -> MemoryBuilder 
(<->) address (s:xs) memory = write8 address s $ (<->) (address + 1) xs memory
(<->) address [] memory = memory

decodeAddress :: Int -> (Int, Int)
decodeAddress address = ( shiftR (address .&. 0xFF00) 8, address .&. 0xFF)

instance Memory ListMemory where
    read8 address memory = read8FromPage (getMemoryPage memory pageNumber) offset where 
        (pageNumber, offset) = decodeAddress address
    read16 address memory= read16FromPage (getMemoryPage memory pageNumber) offset where 
        (pageNumber, offset) = decodeAddress address
    write8 address value memory = updateMemoryPage memory pageNumber $ write8ToPage offset value where
        (pageNumber, offset) = decodeAddress address
    write16 address value memory = updateMemoryPage memory pageNumber $ write16ToPage offset value where
        (pageNumber, offset) = decodeAddress address

updateMemoryPage :: ListMemory -> Int -> ([Int] -> [Int]) -> ListMemory
updateMemoryPage ( (currentPage:restPages)) pageNumber updater
    | pageNumber == 0 =  ((updater currentPage):restPages)
    | otherwise =  (currentPage:t) where
         t = updateMemoryPage ( restPages) (pageNumber - 1) updater 
updateMemoryPage [] pageNumber updater
    | pageNumber == 0 =  [(updater emptyPage)]
    | otherwise = (emptyPage:t) where
         t = updateMemoryPage [] (pageNumber - 1) updater

emptyPage :: [Int]
emptyPage = take 256 (repeat 0)

getMemoryPage :: ListMemory -> Int -> [Int]
getMemoryPage [] _ = emptyPage 
getMemoryPage ( (firstPage:otherPages)) pageNumber 
    | pageNumber == 0 = firstPage
    | otherwise = getMemoryPage ( otherPages) (pageNumber - 1)

read8FromPage :: [Int] -> Int -> Int
read8FromPage page offset = page!!offset

-- TODO: optimize
read16FromPage :: [Int] -> Int -> (Int, Int)
read16FromPage page address = (read8FromPage page (address + 1), read8FromPage page address)

write8ToPage :: Int -> Int -> [Int] -> [Int]
write8ToPage offset value (x:xs) 
    | offset == 0 = value:xs
    | otherwise = x:write8ToPage (offset - 1) value xs
write8ToPage offset value []
    | offset == 0 = [value]
    | otherwise = 0:write8ToPage (offset - 1) value [] 

write16ToPage :: Int -> (Int, Int) -> [Int] -> [Int]
write16ToPage offset (b2, b1) page = write8ToPage (offset + 1) b2 $write8ToPage offset b1 page



src/CPU6502.hs
{-# LANGUAGE TypeSynonymInstances #-}
{-# LANGUAGE FlexibleInstances #-}

module CPU6502 where

import System.Environment
import Data.Bits
import Memory


data CPU6502 = CPU6502 { 
    memory :: ListMemory
    , tick :: Int
    , registerACC:: Int
    , registerPC :: Int
    , pcCache :: [Int]
    , registerS :: Int
    , registerX :: Int
    , registerY :: Int
    , flagC :: Bool
    , flagZ :: Bool
    , flagN :: Bool
    , flagV :: Bool
    , flagI :: Bool
    , flagD :: Bool} deriving Show


cpuFromMemory :: ListMemory -> CPU6502
cpuFromMemory memory = CPU6502 { 
    memory = memory, 
    tick = 0,
    registerACC = 0,
    registerPC = 0x8000,
    pcCache = [],
    registerS = 0,
    registerX = 0,
    registerY = 0,
    flagC = False,
    flagZ = False,
    flagN = False,
    flagV = False,
    flagI = False,
    flagD = False}

newCPU :: MemoryBuilder -> CPU6502
newCPU memoryBuilder = cpuFromMemory $ memoryBuilder []

addTicks :: Int -> CPU6502 -> CPU6502
addTicks tick cpu@CPU6502{tick = oldValue} = cpu{tick = oldValue + tick}

getAccRegister :: CPU6502 -> Int 
getAccRegister = registerACC 
setAccRegister :: Int -> CPU6502 -> CPU6502
setAccRegister value = updateAccRegister (\_ -> value)
updateAccRegister :: (Int -> Int) -> CPU6502 -> CPU6502 
updateAccRegister updateFunc cpu@CPU6502{registerACC = accValue} = cpu { registerACC = updateFunc accValue}

getPCRegister :: CPU6502 -> Int 
getPCRegister = registerPC 
setPCRegister :: Int -> CPU6502 -> CPU6502
setPCRegister value cpu = (clearPCCache.(updatePCRegister (\_ -> value))) cpu
-- TODO: ckear pc cache
updatePCRegister :: (Int -> Int) -> CPU6502 -> CPU6502 
updatePCRegister updateFunc cpu@CPU6502{registerPC = pcValue} = cpu { registerPC = updateFunc pcValue}

clearPCCache :: CPU6502 -> CPU6502
clearPCCache cpu = cpu{pcCache = []}

readNextWord8 :: CPU6502 -> (Int, CPU6502)
readNextWord8 cpu = (value, updatePCRegister (1+) newCPU) where 
    pc = getPCRegister cpu
    (value, newCPU) = readWord8Cached pc (pcCache cpu) cpu
    
readWord8Cached :: Int -> [Int] -> CPU6502 -> (Int, CPU6502)
readWord8Cached address []  cpu = (value, cpu{pcCache=newCache}) where
    (pageNumber, offset) = decodeAddress address
    page = getMemoryPage (memory cpu) pageNumber
    cache = drop offset page
    (value:newCache) = cache
readWord8Cached address cache cpu = (value, cpu{pcCache=newCache}) where
    (value:newCache) = cache

readNextWord16 :: CPU6502 -> (Int, CPU6502)
readNextWord16 cpu = (valueFrom16 ((b2, b1)), cpu2) where 
    (b1, cpu1) = readNextWord8 cpu
    (b2, cpu2) = readNextWord8 cpu1

getXRegister :: CPU6502 -> Int 
getXRegister = registerX
setXRegister :: Int -> CPU6502 -> CPU6502
setXRegister value cpu = updateXRegister (\_ -> value) cpu
updateXRegister :: (Int -> Int) -> CPU6502 -> CPU6502 
updateXRegister updateFunc cpu@CPU6502{registerX = xValue} = cpu { registerX = updateFunc xValue}

getYRegister :: CPU6502 -> Int 
getYRegister = registerY 
setYRegister :: Int -> CPU6502 -> CPU6502
setYRegister value cpu = updateYRegister (\_ -> value) cpu
updateYRegister :: (Int -> Int) -> CPU6502 -> CPU6502 
updateYRegister updateFunc cpu@CPU6502{registerY = yValue} = cpu { registerY = updateFunc yValue}

getFlagC :: CPU6502 -> Bool
getFlagC = flagC
setFlagC :: Bool -> CPU6502 -> CPU6502
setFlagC value cpu = cpu{flagC = value}

getFlagZ :: CPU6502 -> Bool
getFlagZ = flagZ
setFlagZ :: Bool -> CPU6502 -> CPU6502
setFlagZ value cpu = cpu{flagZ = value}

getFlagN :: CPU6502 -> Bool
getFlagN = flagN
setFlagN :: Bool -> CPU6502 -> CPU6502
setFlagN value cpu = cpu{flagN = value}

getFlagV :: CPU6502 -> Bool
getFlagV = flagV
setFlagV :: Bool -> CPU6502 -> CPU6502
setFlagV value cpu = cpu{flagV = value}

getFlagI :: CPU6502 -> Bool
getFlagI = flagI
setFlagI :: Bool -> CPU6502 -> CPU6502
setFlagI value cpu = cpu{flagI = value}

getFlagD :: CPU6502 -> Bool
getFlagD = flagV
setFlagD :: Bool -> CPU6502 -> CPU6502
setFlagD value cpu = cpu{flagD = value}

instance Memory CPU6502 where
    read8 address cpu  = read8 address (memory cpu) 
    read16 address cpu  = read16 address (memory cpu) 
    write8 address value cpu = cpu { memory = write8 address value (memory cpu)}
    write16 address value cpu = cpu { memory = write16 address value (memory cpu)} 

type CPUOperation = CPU6502 -> CPU6502 

resetCPU :: CPUOperation 
resetCPU  = cpuFromMemory.memory 

iterateProcessor :: CPUOperation 
iterateProcessor cpu = decodeNextOperation command newCPU
    where (command, newCPU) = readNextWord8 cpu

decodeNextOperation :: Int -> CPUOperation
decodeNextOperation command = 
    case command of
        0x65 -> adc.argZP.(addTicks 3)
        0x75 -> adc.argZPX.(addTicks 4)
        0x6D -> adc.argABS.(addTicks 4)
        0x7D -> adc.argABSX.(addTicks 4)
        0x79 -> adc.argABSY.(addTicks 4)
        0x69 -> adc.argIMM.(addTicks 2)
        0x61 -> adc.argINDX.(addTicks 6)
        0x71 -> adc.argINDY.(addTicks 5)
        0xEA -> nop.argNOP.(addTicks 2)
        0xE8 -> inx.argNOP.(addTicks 2)
        0xCA -> dex.argNOP.(addTicks 2)
        0xC8 -> iny.argNOP.(addTicks 2)
        0x88 -> dey.argNOP.(addTicks 2)
        0x90 -> bcc.argREL.(addTicks 2)
        0xB0 -> bcs.argREL.(addTicks 2)
        0xF0 -> beq.argREL.(addTicks 2)
        0x30 -> bmi.argREL.(addTicks 2)
        0xD0 -> bne.argREL.(addTicks 2)
        0x10 -> bpl.argREL.(addTicks 2)
        0x50 -> bvc.argREL.(addTicks 2)
        0x70 -> bvs.argREL.(addTicks 2)
        0x18 -> clc.argNOP.(addTicks 2)
        0xD8 -> cld.argNOP.(addTicks 2)
        0x58 -> cli.argNOP.(addTicks 2)
        0xB8 -> clv.argNOP.(addTicks 2)

        0x85 -> sta.argZP.(addTicks 3)
        0x95 -> sta.argZPX.(addTicks 4)
        0x8D -> sta.argABS.(addTicks 4)
        0x9D -> sta.argABSX.(addTicks 5)
        0x99 -> sta.argABSY.(addTicks 5)

data MemoryAccessor = MemoryAccessor Int CPU6502 | ConstMemoryAccessor Int | NOPMemoryAccessor
readValue :: MemoryAccessor -> Int
readValue (ConstMemoryAccessor value) = value 
readValue (MemoryAccessor address cpu) = read8 address cpu
writeValue :: MemoryAccessor -> Int -> CPU6502
writeValue (MemoryAccessor address cpu) value = write8 address value cpu

instance Show MemoryAccessor where
    show = show.readValue 


argNOP :: CPU6502 -> (MemoryAccessor, CPU6502)
argNOP cpu = (NOPMemoryAccessor, cpu)

argZP :: CPU6502 -> (MemoryAccessor, CPU6502)
argZP cpu = ( MemoryAccessor arg newCPU, newCPU) where
    (arg, newCPU) = readNextWord8 cpu

argZPX :: CPU6502 -> (MemoryAccessor, CPU6502)
argZPX cpu = ( MemoryAccessor argAddress newCPU, newCPU) where
    argAddress = getXRegister newCPU  + arg
    (arg, newCPU) = readNextWord8 cpu

argABS :: CPU6502 -> (MemoryAccessor, CPU6502)
argABS cpu = ( MemoryAccessor arg newCPU, newCPU) where
    (arg, newCPU) = readNextWord16 cpu

argABSX :: CPU6502 -> (MemoryAccessor, CPU6502)
argABSX cpu = ( MemoryAccessor argAddress newCPU, newCPU) where
    argAddress =  getXRegister newCPU + arg
    (arg, newCPU) = readNextWord16 cpu

argABSY :: CPU6502 -> (MemoryAccessor, CPU6502)
argABSY cpu = ( MemoryAccessor argAddress newCPU, newCPU) where
    argAddress =  getYRegister newCPU + arg 
    (arg, newCPU) = readNextWord16 cpu

argINDX :: CPU6502 -> (MemoryAccessor, CPU6502)
argINDX cpu = ( MemoryAccessor argAddress newCPU, newCPU) where
    indAddress = getXRegister newCPU + arg
    argAddress = valueFrom16 $ read16 indAddress newCPU 
    (arg, newCPU) = readNextWord8 cpu

argINDY :: CPU6502 -> (MemoryAccessor, CPU6502)
argINDY cpu = ( MemoryAccessor argAddress newCPU, newCPU) where
    argAddress = getYRegister newCPU +  (valueFrom16 $ read16 arg newCPU) 
    (arg, newCPU) = readNextWord8 cpu

argIMM :: CPU6502 -> (MemoryAccessor, CPU6502)
argIMM cpu = (ConstMemoryAccessor arg, newCPU) where
    (arg, newCPU) = readNextWord8 cpu

argREL :: CPU6502 -> (MemoryAccessor, CPU6502)
argREL cpu = (ConstMemoryAccessor (toSignedByte arg), newCPU) where
    (arg, newCPU) = readNextWord8 cpu

toSignedByte :: Int -> Int
toSignedByte x 
    | x < 128 = x
    | otherwise = x - 0xFF - 1

valueFrom16 :: (Int, Int) -> Int
valueFrom16 (b2, b1) = b2*2^8 + b1

adc :: (MemoryAccessor, CPU6502) -> CPU6502 
adc (arg, cpu) = 
     ((setFlagC newCarryFlagValue)
     .(setFlagZ newZeroFlagValue)
     .(setFlagN newSignFlagValue)
     .(setFlagV newOverflowFlagValue)
     .(setAccRegister result)) cpu where
        accValue = getAccRegister cpu
        argValue = readValue arg
        carryFlagValue = getFlagC cpu
        sum = accValue + argValue + if carryFlagValue then 1 else 0
        newCarryFlagValue = if sum > 0xFF then True else False
        newZeroFlagValue = if result == 0 then True else False
        newSignFlagValue = if result > 0x7F then True else False
        newOverflowFlagValue = if (((accValue `xor`argValue) .&. 0x80) == 0) && (((accValue `xor` result) .&. 0x80) /= 0) then True else False
        result  = if newCarryFlagValue then sum - 0x100 else sum

bcc :: (MemoryAccessor, CPU6502) -> CPU6502
bcc (arg, cpu) = if (getFlagC cpu) == False then setPCRegister ((getPCRegister cpu) + (readValue arg))  cpu else cpu

bcs :: (MemoryAccessor, CPU6502) -> CPU6502
bcs (arg, cpu) = if (getFlagC cpu) == True then setPCRegister ((getPCRegister cpu) + (readValue arg))  cpu else cpu

beq :: (MemoryAccessor, CPU6502) -> CPU6502
beq (arg, cpu) = if (getFlagZ cpu) == True then setPCRegister ((getPCRegister cpu) + (readValue arg))  cpu else cpu

bne :: (MemoryAccessor, CPU6502) -> CPU6502
bne (arg, cpu) = if (getFlagZ cpu) == False then setPCRegister ((getPCRegister cpu) + (readValue arg))  cpu else cpu

bmi :: (MemoryAccessor, CPU6502) -> CPU6502
bmi (arg, cpu) = if (getFlagN cpu) == True then setPCRegister ((getPCRegister cpu) + (readValue arg))  cpu else cpu

bpl :: (MemoryAccessor, CPU6502) -> CPU6502
bpl (arg, cpu) = if (getFlagN cpu) == False then setPCRegister ((getPCRegister cpu) + (readValue arg))  cpu else cpu

bvs :: (MemoryAccessor, CPU6502) -> CPU6502
bvs (arg, cpu) = if (getFlagV cpu) == True then setPCRegister ((getPCRegister cpu) + (readValue arg))  cpu else cpu

bvc :: (MemoryAccessor, CPU6502) -> CPU6502
bvc (arg, cpu) = if (getFlagV cpu) == False then setPCRegister ((getPCRegister cpu) + (readValue arg))  cpu else cpu

clc :: (MemoryAccessor, CPU6502) -> CPU6502
clc (_, cpu) = setFlagC False cpu

cld :: (MemoryAccessor, CPU6502) -> CPU6502
cld (_, cpu) = setFlagD False cpu

cli :: (MemoryAccessor, CPU6502) -> CPU6502
cli (_, cpu) = setFlagI False cpu

clv :: (MemoryAccessor, CPU6502) -> CPU6502
clv (_, cpu) = setFlagV False cpu

nop :: (MemoryAccessor, CPU6502) -> CPU6502 
nop (_, cpu) = cpu

inx :: (MemoryAccessor, CPU6502) -> CPU6502 
inx (_, cpu) = updateXRegister (1+) cpu

dex :: (MemoryAccessor, CPU6502) -> CPU6502 
dex (_, cpu) = updateXRegister (1-) cpu

iny :: (MemoryAccessor, CPU6502) -> CPU6502 
iny (_, cpu) = updateYRegister (1+) cpu

dey :: (MemoryAccessor, CPU6502) -> CPU6502 
dey (_, cpu) = updateYRegister (1-) cpu

-- S
sta :: (MemoryAccessor, CPU6502) -> CPU6502
sta (arg, cpu) = writeValue arg $ getAccRegister cpu
